### ISSUE TITLE
[SUGGESTION] Statement-expressions, `result` vs `return`

### ISSUE THREAD

**Issue by msadeqhe** [hsutter/cppfront#391]
### Preface

Statement-expression is a solution to have a group of statements in place of an expression.

I suggest to have statement-expressions in Cpp2 with syntax `: = { ... }` and the result of them are returned with `result` keyword. In this way, the following code:

```cpp
call(: = {
    n: = num();
    result n + 0;
}, : = {
    n: = num();
    result n + 1;
});
```

... is somehow equal to the following code except that we don't have to specify the type of arguments:

```cpp
arg0: int;
{
    n: = num();
    arg0 = n + 0;
}
arg1: int;
{
    n: = num();
    arg1 = n + 1;
}
call(arg0, arg1);
```

Statement-expressions can be used to declare variables too:

```cpp
// The type of `variable` is `int`.
variable: = {
    value: int;
    //: statements...
    result value;
}
```

### Suggestion Detail

Currently we declare functions in Cpp2 in the following syntax:

```cpp
// This is the normal style.
func0: (param) = {
    return param;
}

// This is the expression style.
func1: (param) = param;
```

In a similar syntax, we can declare variables in Cpp2 with statement-expressions:

```cpp
// This is statement-expression style.
var0: = {
    result 0;
}

// This is the normal style.
var1: = 0;
```

Statement-expressions are unnamed variables in a similar manner that lambdas are unnamed functions:

- We have to use `result` instead of `return` within statement-expressions.
- We can use a statement-expression everywhere that an expression is required.
- A declaration within their `{ ... }` has the lifetime of that scope.
- Their `{ ... }` will be executed one time in their place.
- They don't create a new function scope, therefore variables from outer scope don't have to be captured within their `{ ... }`.

Statement-expressions will be grammatically incorrect to be used in place of statements. Also it's an error to have an unused value in Cpp2:

```
// ERROR! It has to be in place of an expression not an statement.
// ERROR! Also It's an unused value.
: = {
    x: = 0;
    result x + 1;
}

// OK. This works.
{
    x: = 0;
    result x + 1;
}
```

### Expressions, Functions and Blocks

Now `{ ... }` has different meanings in the following categories:

1. `: Type = { ... }` is a **statement-expression** in which `Type` can be omitted and deduced from `{ ... }`.
    - Its `{ ... }` ends with `result something;`.
    - It can be a part of variable declaration, or in place of an expression (e.g. function argument):
    ```cpp
    x0: = 0;
    x1: = { result 0; }
    call(: = { result 0; })
    ```
2. `: (params) -> Type = { ... }` is a **function** or a **lambda** in which `-> Type` can be omitted if the return type is `void`.
    - Its `{ ... }` ends with `return something;` or the end of `}`.
    - It can be a part of function declaration, or in place of an expression (e.g. function argument):
    ```cpp
    x0: (param) = param;
    x1: (param) = { return param; }
    call(: (param) = { return param; })
    ```
3. `{ ... }` is a **statement block**.
    - Its `{ ... }` ends with:
        - ... the end of `}`.
        - ... `break` and `continue` in loop control structures.
        - ... `result something;` when their outer scope is a statement-expression.
        - ... `return something;` when their outer scope is a function or a lambda.
    - It can be a part of control structure:
    ```cpp
    if condition { call(); }
    ```
4. `(params) { ... }` is a **parameterized statement block**.
    - Its `{ ... }` ends with:
        - ... the end of `}`.
        - ... `break` and `continue` in loop control structures.
        - ... `result something;` when their outer scope is a statement-expression.
        - ... `return something;` when their outer scope is a function or a lambda.
    - It can be a part of control structure:
    ```cpp
    for items do (item) { call(item); }
    ```

In all of the above categories, they have this similarity:

- A declaration within their `{ ... }` has the lifetime of that scope.

Additinally they have different behaviours in statement execution and variable capturing:

- **Statement-expressions** (case 1) have to execute their `{ ... }` one time in their place.
    - But other cases can execute their `{ ... }` multiple times in any place.
- **Functions** and **lambdas** (case 2) create a new function scope therefore variables from outer scope have to be captured within their `{ ... }`.
    - But other cases don't create a new function scope, therefore variables from outer scope don't have to be captured within their `{ ... }`.

I should mention that statement-expressions, functions and lambdas, statement blocks and parameterized statement blocks can be nested inside each other.

### Your Questions

**Will your feature suggestion eliminate X% of security vulnerabilities of a given kind in current C++ code?**

Yes. In a way that it helps to organize code and separate related statements in nested blocks to easily fix or prevent unwanted bugs which are a result of having mixed unrelated statements together.

**Will your feature suggestion _automate or eliminate_ X% of current C++ guidance literature?**

Yes. In the following ways:

- It will make the declaration syntax of variables and functions to be similar and consistent.
- It helps programmers to separate related statements easily without defining extra variables.
- It allows programmers to use `if`, `while` and other control structures as an expression which is more readable in variable declaration, see the example below.
- It allows programmers to be less specific about types in generic programming, see the example below.

Consider how the following code:

```cpp
variable: = {
    if condition {
        result hello();
    }
    else {
        result bye();
    }
}
```

... is more readable and generic than the following code which we have to also specify the type of the variable:

```cpp
variable: int;

if condition {
    variable: = hello();
}
else {
    variable: = bye();
}
```

### Considered Alternatives

Instead of `result` keyword, it's possible to use symbols such as `=>` or nothing (like in Rust). For example:

```cpp
variable: = {
    if condition {
        => hello();
    }
    else {
        => bye();
    }
}
```

... or this one (like in Rust):

```cpp
variable: = {
    if condition {
        hello();
    }
    else {
        bye();
    }
}
```

Howerver I think having a keyword like `result` or a notation like `=>` is more readable than having nothing.

### Additinal Suggestion

It's not a part of my main suggestion. Additionally I suggest to consider to require `do` before `{ ... }` and call it `do { ... }` language construct:

```cpp
variable: = 0;

do {
    variable: = 0;
}
```

It's useful because after [this commit](https://github.com/hsutter/cppfront/commit/928186e4db6f7c3fba7b8df8b4ff9648e1c938e9), Cpp2 has statement scope parameters:

```cpp
(x: = 0) for items do (item) {
    //: statements...
}
```

`(x: = 0)` is different from `(item)` and it's obvious from the example. `(x: = 0)` is for initialization and `(item)` is for argument passing.

Also for `do { ... } while condition;` we would have:

```cpp
(x: = 0) do {
    //: statements...
}
while condition;
```

Again it's obvious that `(x: = 0)` is for initialization, and nothing is declared for argument passing.

But in the following example:

```cpp
(x: = 0) {
    //: statements...
}
```

`(x: = 0)` is for initialization but it feels like its syntax is for argument passing like in `for args do (arg) { ... }`, because the parenthesis are immediately followed by braces like `(...) { ... }`. On the other hand if it was written with `do { ... }`, then it would be like this:

```cpp
(x: = 0) do {
    //: statements...
}
```

... which is also obvious that `(x: = 0)` is for initialization because it's not immediately followed by braces.

So in a nutshell with `do { ... }`:

- It will be obvious if `(...)` is for initialization or for argument passing, the syntax will be generally like this:
    - `(inits) construct (params) { ... }` in which `construct` may be `if`, `while`, `do`, `for` or anything else in the future.
- All statement blocks will be a part of control structures.
- `{ ... }` creates a new scope for variables and declarations. It's important enough to deserve a unique name and a control structure.
- `do { ... }` will be complementary to `do { ... } while condition;`.
- If Cpp2 use `{ ... }` for initializer lists or other type of lists, then `do { ... }` won't be ambiguous with them.

### References

This suggestion is inspired from discussions in [this issue](https://github.com/hsutter/cppfront/issues/386) and [this paper](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2806r1.html) mentioned by @JohelEGP and [this informative comment](https://github.com/hsutter/cppfront/issues/386#issuecomment-1518194517) from @hsutter which describes _Unifying Functions and Blocks_.

**Comment by msadeqhe** [comment]
> So in a nutshell with `do { ... }`:
> 
> * It will be obvious if `(...)` is for initialization or for argument passing in parameterized statement blocks, the syntax will be generally like this:
>   
>   * `(inits) construct (params) { ... }` in which `construct` may be `if`, `while`, `do`, `for` or anything else in the future.
> * All statement blocks will be a part of control structures.
> * `{ ... }` creates a new scope for variables and declarations. It's important enough that it deserves a control structure with a name.
> * `do { ... }` will be complementary to `do { ... } while condition;`.
> * If Cpp2 use `{ ... }` for initializer lists or other type of lists, then `do { ... }` won't be ambiguous with them.

I'm going to express the opposite view that why `{ ... }` is better than `do { ... }` in response to my _Additional Suggestion_:

- It's obvious enough that `(...)` is for initialization or for argument passing in parameterized statement blocks, the syntax will be generally like this (in the following list: `keywords` may be `if`, `while`, `do`, `for`, `next` or anything else in the future, `(inits)` is for initialization and `(params)` is for argument passing in parameterized statement blocks):
    - `(inits) keywords (params) { ... }`
    - `(inits) { ... }`
    - `(inits) keywords { ... }`
    - `keywords (params) { ... }`
    - In a nutshell, `(params)` are after `keywords` and `(inits)` are before them. Also when there is only one `(...)` it will be `(inits)`.
- Statement blocks can be a part of control structures, or can be used alone without any control structures.
- `{ ... }` creates a new scope for variables and declarations. It's obvious from `{ ... }` notation and nothing more is needed.
- `{ ... }` is already a part of statement-expressions, functions and lambdas. When they are alone without any control structures, they are a statement block or a parameterized statement block.

**Comment by AbhinavK00** [comment]
If I understand correctly, his is just `do expressions` paper from Barry Revzin (without the `do` part ofcourse), no?
I like this suggestion, helps to get rid of immediately invoked lambda expressions (which are not bad, but this is simpler).

**Comment by msadeqhe** [comment]
Yes, that's it. Only the syntax is changed to `: = { ... result something; ... }` for Cpp2.

> I like this suggestion, helps to get rid of immediately invoked lambda expressions (which are not bad, but this is simpler).

Also capturing variable, is not needed.
